RearrangeCarGroup8Movements.py

```diff
@@ -0,0 +1,195 @@
+        self.empty_position = car_to_move_position
+
+    def _get_not_arranged_car_index(self):
+
+        '''Finds first car index not yet standing on the final destination
+        '''
+
+        car = random.choice(tuple(self.not_arranged_cars)) # O(1)
+        return self.car_position_dict[car]
+
+    def _rearrange_one_car(self):
+
+        '''Puts to the current empty space the car
+        that should be on this place according to the final arrangement
+        if current empty space position is different from the final order empty
+        space position
+
+        Otherwise picks up the random not arranged car and places it to the empty space
+        in this case the number of not arranged cars stays the same
+
+        Complexity is O(1)
+        '''
+
+        if self.final_car_order[self.empty_position] == 0:
+            car_to_move_position = self._get_not_arranged_car_index()
+            car_to_move = self.current_car_order[car_to_move_position]
+        else:
+            car_to_move = self.final_car_order[self.empty_position]
+            car_to_move_position = self.car_position_dict[car_to_move]
+            self.not_arranged_cars.difference_update([car_to_move]) # O(1)
+
+
+        movement = Movement(car_to_move, car_to_move_position, self.empty_position)
+        if self.print_console:
+            print movement
+        else:
+            movement.print_log_file()
+
+        self._move_car_to_empty_slot(car_to_move_position)
+
+    def _reset_variables_to_init_state(self):
+
+        '''Resets variables to the same state when class was initialized
+        Needed for running rearrange_all_cars function again
+        '''
+
+        self.current_car_order = self.initial_car_order
+        self.empty_position = self.current_car_order.index(0)
+        self.car_position_dict = {car:pos for pos, car in \
+                                  enumerate(self.initial_car_order) if car} # time O(n)
+        self.not_arranged_cars = {car for i, car in enumerate(self.initial_car_order)\
+                                  if self.initial_car_order[i] != self.final_car_order[i]\
+                                  and car} # time O(n)
+
+    def rearrange_all_cars(self):
+
+        '''Rearranges all cars and prints the moves
+        Note 1: In case there are more than 1000 moves to print
+        logging into the file is done
+        Note 2: Possible to run the function many times
+
+        '''
+
+        if self.initial_car_order == self.final_car_order:
+            assert ValueError, 'Cars are already in the final order, no moves are required'
+        elif self.current_car_order == self.final_car_order:
+            self._reset_variables_to_init_state()
+
+        if len(self.not_arranged_cars) > 1000:
+            self._create_log_file()
+            self.print_console = False
+
+        '''Final solution is O(n)
+        '''
+
+        while self.not_arranged_cars:
+            self._rearrange_one_car() # O(1)
+

```

Comment:
What about moving `self._reset_variables_to_init_state()` to the very beginning of this function, so that we it automatically reset all states before checking not arranged cars? (Then we can remove the first `if` and `elif` checks. Since if the init state is equal to the final state, there will be no cars in `self.not_arranged_cars` and en empty list will be returned.) 

And then since we plan to return a movement object from `self._rearrange_one_car()`, we can let this function return a list of movements, and write another function that does the printing of the list?